csharp-generator/src/csharp-server.ts

```diff
@@ -0,0 +1,112 @@
+import { AstRoot, astToJson, OptionalType, VoidPrimitiveType } from "@sdkgen/parser";
+import { decodeType, encodeType, generateEnum, generateStruct, generateTypeName, ident } from "./helpers";
+
+interface Options {}
+
+export function generateCSharpServerSource(ast: AstRoot, options: Options) {
+    let code = `using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Numerics;
+using System.Text.Json;
+using System.Threading.Tasks;
+using SdkgenRuntime;
+
+namespace SdkgenGenerated
+{
+    public abstract class Api : BaseApi
+    {`;
+    for (const op of ast.operations) {
+        const returnTypeAngle = op.returnType instanceof VoidPrimitiveType ? "" : `<${generateTypeName(op.returnType)}>`;
+        code += `
+        virtual public Task${returnTypeAngle} ${op.name}(${[
+            "Context ctx",
+            ...op.args.map(arg => `${generateTypeName(arg.type)} ${ident(arg.name)}`),
+        ].join(", ")})
+        {
+            return Task.FromException${returnTypeAngle}(new FatalException("function '${op.name}' not implemented"));
+        }
+`;
+    }
+    code += `
+        public async Task ExecuteFunction(Context context_, Utf8JsonWriter resultWriter_)
+        {
+            switch (context_.Name)
+            {`;
+    for (const op of ast.operations) {
+        code += `
+                case ${JSON.stringify(op.name)}:
+                    {`;
+
+        for (const arg of op.args) {
+            code += `
+                        JsonElement ${arg.name}Json_;
+                        if (!context_.Args.TryGetValue(${JSON.stringify(arg.name)}, out ${arg.name}Json_))
+                        {
+                            ${
+                                arg.type instanceof OptionalType
+                                    ? `${arg.name}Json_ = new JsonElement()`
+                                    : `throw new FatalException("'${op.name}().args.${arg.name}' must be set to a value of type ${arg.type.name}");`
```

Comment:
Should also throw same errors as Node

https://github.com/sdkgen/sdkgen/blob/1e431aa5403c5aa8a1e35c761cf5a333186074bc/node-runtime/src/encode-decode.ts#L53